File: src/main/java/org/auscope/portal/server/eavl/ParameterDetails.java
package org.auscope.portal.server.eavl;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;

public class ParameterDetails implements Serializable {
    /** Unique name of this parameter */
    private String name;
    /** Total number of numeric values in this parameter*/
    private int totalNumeric;
    /** Total number of missing values in this parameter (null or whitespace only strings)*/
    private int totalMissing;
    /** Total number of values that are NOT parseable as a number and are non empty*/
    private int totalText;
    /** Total number of numerical values that equal 0 (exactly) */
    private int totalZeroes;
    /** A map of every non numeric value (including null string) and their associated counts.*/
    private Map<String, Integer> textValues;
    /** The index of the column in the CSV file (0 based)*/
    private int columnIndex;
    /** The maximum value in this column (ignoring 0 values)*/
    private double maxValue;
    /** The minimum value in this column (ignoring 0 values)*/
    private double minValue;

    /**
     * @param name Unique name of this parameter
     * @param columnIndex The index of the column in the CSV file (0 based)
     */
    public ParameterDetails(String name, int columnIndex) {
        this.name = name;
        this.columnIndex = columnIndex;
        this.textValues = new HashMap<String, Integer>();
        this.maxValue = -Double.MAX_VALUE;
        this.minValue = Double.MAX_VALUE;
    }

    /**
     * Unique name of this parameter
     * @return
     */
    public String getName() {
        return name;
    }
    /**
     * Unique name of this parameter
     * @param name
     */
    public void setName(String name) {
        this.name = name;
    }
    /**
     * Total number of numeric values in this parameter
     * @return
     */
    public int getTotalNumeric() {
        return totalNumeric;
    }
    /**
     * Total number of numeric values in this parameter
     * @param totalNumeric
     */
    public void setTotalNumeric(int totalNumeric) {
        this.totalNumeric = totalNumeric;
    }
    /**
     * Total number of missing values in this parameter (null or whitespace only strings)
     * @return
     */
    public int getTotalMissing() {
        return totalMissing;
    }
    /**
     * Total number of missing values in this parameter (null or whitespace only strings)
     * @param totalMissing
     */
    public void setTotalMissing(int totalMissing) {
        this.totalMissing = totalMissing;
    }
    /**
     * Total number of values that are NOT parseable as a number and are non empty
     * @return
     */
    public int getTotalText() {
        return totalText;
    }
    /**
     * Total number of values that are NOT parseable as a number and are non empty
     * @param totalText
     */
    public void setTotalText(int totalText) {
        this.totalText = totalText;
    }

    /**
     * Total number of numerical values that equal 0 (exactly)
     * @return
     */
    public int getTotalZeroes() {
        return totalZeroes;
    }

    /**
     * Total number of numerical values that equal 0 (exactly)
     * @param totalZeroes
     */
    public void setTotalZeroes(int totalZeroes) {
        this.totalZeroes = totalZeroes;
    }

    /**
     * A map of every non numeric value (including null string) and their associated counts.
     * @return
     */
    public Map<String, Integer> getTextValues() {
        return textValues;
    }
    /**
     * A map of every non numeric value (including null string) and their associated counts.
     * @param textValues
     */
    public void setTextValues(Map<String, Integer> textValues) {
        this.textValues = textValues;
    }

    /** The index of the column in the CSV file*/
    public int getColumnIndex() {
        return columnIndex;
    }

    /** The index of the column in the CSV file*/
    public void setColumnIndex(int columnIndex) {
        this.columnIndex = columnIndex;
    }

    /**
     * The maximum value in this column (ignoring 0 values)
     * @return
     */
    public double getMaxValue() {
        return maxValue;
    }

    /**
     * The maximum value in this column (ignoring 0 values)
     * @param maxValue
     */
    public void setMaxValue(double maxValue) {
        this.maxValue = maxValue;
    }

    /**
     * The minimum value in this column (ignoring 0 values)
     * @return
     */
    public double getMinValue() {
        return minValue;
    }

    /**
     * The minimum value in this column (ignoring 0 values)
     * @param minValue
     */
    public void setMinValue(double minValue) {
        this.minValue = minValue;
    }

    @Override
    public String toString() {
        return "ParameterDetails [name=" + name + "]";
    }
}

File: src/main/webapp/js/eavl/models/ParameterDetails.js
/**
 * A ParameterDetails is a simple set of statistics about a single parameter (column)
 * in a CSV dataset
 */
Ext.define('eavl.models.ParameterDetails', {
    extend: 'Ext.data.Model',

    statics : {
        UOM_PPM : 'ppm',
        UOM_PCT : 'pct',
        
        STATUS_GOOD : 0,
        STATUS_WARNING : 1,
        STATUS_ERROR : 2,

        /**
         * Compares two parameter details based on "status" and then on column index
         *
         * @param a eavl.models.ParameterDetails First parameter to compare
         * @param b eavl.models.ParameterDetails Second parameter to compare
         */
        sortSeverityFn : function(a, b) {
            var aStatus = a.calculateStatus();
            var bStatus = b.calculateStatus();

            if (aStatus === bStatus) {
                return a.get('columnIndex') - b.get('columnIndex');
            }

            return bStatus - aStatus;
        },

        /**
         * Compares two parameter details based on "columnIndex"
         *
         * @param a eavl.models.ParameterDetails First parameter to compare
         * @param b eavl.models.ParameterDetails Second parameter to compare
         */
        sortColIndexFn : function(a, b) {
            return a.get('columnIndex') - b.get('columnIndex');
        },

        /**
         * Given an array of eavl.models.ParameterDetails, find an instance with matching name
         * and remove it from the array. Return the removed element (or null) if it DNE.
         *
         * @param pdArray eavl.models.ParameterDetails[] Will have the first matching instance removed
         * @param name String The name to search fo
         * @param dontDelete if set to true pdArray will not be modifiedr
         */
        extractFromArray : function(pdArray, name, dontDelete) {
            var matchingPd = null;
            var index = -1;
            Ext.each(pdArray, function(pd, i) {
                if (pd.get('name') === name) {
                    index = i;
                    matchingPd = pd;
                    return false;
                }
            });

            if (index < 0) {
                return null;
            }

            if (!dontDelete) {
                Ext.Array.erase(pdArray, index, 1);
            }
            return matchingPd;
        },

        /**
         * Attempts to convert a string name to a unit of measure guess. Returns null if the name
         * cannot be reliably converted into a uom
         * 
         * @param name String based name to look for a uom identifier
         */
        nameToUom : function(name) {
            var lowerName = name.toLowerCase();
            if (lowerName.indexOf('ppm') >= 0) {
                return eavl.models.ParameterDetails.UOM_PPM;
            } else if (lowerName.indexOf('pct') >= 0 || lowerName.indexOf('percent')  >= 0) {
                return eavl.models.ParameterDetails.UOM_PCT;
            }
            return null;
        }
    },

    fields: [
        { name: 'name', type: 'string' }, //(ideally) unique name of this details
        { name: 'totalNumeric', type: 'int' }, //Total number of numeric values in this parameter
        { name: 'totalMissing', type: 'int' }, //Total number of missing values in this parameter
        { name: 'totalText', type: 'int' }, //Total number of text values in this parameter
        { name: 'totalZeroes', type: 'int' }, //Total number of numerical 0 values in this parameter
        { name: 'columnIndex', type: 'int' }, //The index of the column in the CSV file (0 based)
        { name: 'maxValue', type: 'number' }, //The maximum (non zero) value in this parameter
        { name: 'minValue', type: 'number' }, //The minimum (non zero) value in this parameter
        { name: 'textValues', type: 'auto' }, //Hashmap of totals keyed by each text (non numeric) value in this parameter
        { name: 'scaleFactor', type: 'number'}, //Will not be sent by the backend, but will be entered by the user
        { name: 'displayName', type: 'string', convert: function(v, data) { //The name of this ParameterDetails with any edits applied by the user
            if (!v) {
                return data.get('name');
            }
            return v;
        }},
        { name: 'uom', type: 'string', convert: function(v, data) { //The unit of measure (or null)
            if (v) {
                return v;
            }
            
            //Generate the uom from the layer name by guessing
            return eavl.models.ParameterDetails.nameToUom(data.get('name'));
        }} 

    ],

    idProperty : 'name',

    /**
     * Calculates one of [STATUS_GOOD, STATUS_WARNING, STATUS_ERROR] for this ParameterDetails object.
     *
     */
    calculateStatus : function() {

        if (this.get('totalText') > 0) {
            return eavl.models.ParameterDetails.STATUS_ERROR;
        }

        if (this.get('totalZeroes') > 0) {
            return eavl.models.ParameterDetails.STATUS_ERROR;
        }
        
        if (this.get('uom') !== eavl.models.ParameterDetails.UOM_PPM) {
            return eavl.models.ParameterDetails.STATUS_ERROR;
        }

        var totalDataPoints = this.get('totalNumeric') + this.get('totalMissing');
        var percentageNumeric = (this.get('totalNumeric') * 100) / totalDataPoints;

        if (percentageNumeric < 70) {
            return eavl.models.ParameterDetails.STATUS_WARNING;
        }

        return eavl.models.ParameterDetails.STATUS_GOOD;
    },
    
    /**
     * Copies across all "additional" parameters from pd into this instance.
     * 
     * Additional parameters are the parameters NOT sent by the backend and are
     * instead modified by the user.
     */
    mergeAdditionalParams : function(pd) {
        this.set({
            uom: pd.get('uom'),
            scaleFactor: pd.get('scaleFactor'),
            displayName: pd.get('displayName')
        });
    }
});
